README.md
# Apportion

## An Elixir implementation for apportionment

### Description
Apportionment is most often associated with the [US Congressional apportionment][congress_ep],
using the **Equal Proportions** (or Huntington-Hill) method.

In general, apportionment involves the distribution of a scarce resource among recipients, according
to a pre-determined set of recipient weights. Further, a series of apportionments might occur over
time, with each new apportionment overlaying the previous.

Different domains might make use of apportionment. One example is the allocation (or distribution)
of sellable inventory among a number of retail stores, according to the expected sales rate for
each store. For this usage, recipients would likely begin an apportionment with varying quantities
of existing inventory.

Other apportionment methods exist, notably the [**Quota**][ams_balinski] method,
that address issues with **Equal Proportions**.

## Installation

Add `apportion` to your list of dependencies in `mix.exs`:

  ```elixir
  def deps do
    [{:apportion, "~> 0.0"}]
  end
  ```

## Usage

### Apportion.distribute(weights, quantity, options = %{})

Returns a map (a key for each recipient) with the quantity distributed among recipients
(as values)

#### Parameters

`weights` (required) a [Map] of relative integer proportions for each recipient

`quantity` (required) an [Integer] quantity to apportion

`options` (optional) [Map] keys

* **:prior_portions** a [Map] prior portions for each recipient from a previous apportionment
* **:required_minimum** an [Integer] smallest portion for each recipient

## License

The package is available as open source under the terms of the [MIT License][license].

[congress_ep]: https://en.wikipedia.org/wiki/United_States_congressional_apportionment#The_method_of_equal_proportions
[ams_balinski]: https://www.maa.org/sites/default/files/pdf/upload_library/22/Ford/BalinskiYoung.pdf
[license]: http://opensource.org/licenses/MIT


CHANGELOG.md
## Changelog

### v0.0.1 (2016-09-25)

* initial release


test/apportion_test.exs
defmodule ApportionTest do
  use ExUnit.Case
  doctest Apportion

  test "the truth" do
    assert 1 + 1 == 2
  end
end


test/apportion/util_test.exs
defmodule ApportionUtilTest do
  use ExUnit.Case, async: true
  alias Apportion.Util
  doctest Util

  test "file_read_json_to_map/2" do
    map = %{a: 9061, b: 7179, c: 5259, d: 3319, e: 1182}
    assert map == Util.file_read_json_to_map("test/fixtures/balinski", "table_9_populations.json")
  end

  test "map_values_sum/1" do
    assert 13 == Util.map_values_sum(%{a: 8, b: 2, c: 3})
  end
end


test/apportion/setup_test.exs
defmodule SetupTest do
  use ExUnit.Case, async: true
  alias Apportion.Setup
  doctest Setup

  @keys [:a, :b, :c]

  test "initial_portions/2 w/o options" do
    assert %{a: 0, b: 0, c: 0} == Setup.initial_portions(@keys, [])
  end

  test "initial_portions/2 w required_minimum 2" do
    assert %{a: 2, b: 2, c: 2} == Setup.initial_portions(@keys, required_minimum: 2)
  end

  test "initial_portions/2 w prior_portions 16" do
    portions = %{a: 8, b: 3, c: 5}
    assert portions == Setup.initial_portions(@keys, prior_portions: portions)
  end

  test "initial_portions/2 w :b minimum > prior" do
    portions = %{a: 5, b: 0, c: 1}
    assert %{a: 5, b: 2, c: 2} ==
      Setup.initial_portions(@keys, [
        prior_portions: portions,
        required_minimum: 2
      ])
  end

  test "initial_portions/2 w minimum < prior" do
    portions = %{a: 8, b: 3, c: 5}
    assert portions ==
      Setup.initial_portions(@keys, [
        prior_portions: portions,
        required_minimum: 2
      ])
  end
end


mix.exs
defmodule Apportion.Mixfile do
  use Mix.Project

  def project do
    [
      app: :apportion,
      version: "0.0.1",
      elixir: "~> 1.3",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      package: package(),
      description: "Elixir implementation for apportionment",
      test_coverage: [tool: ExCoveralls]
    ]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [applications: [:logger]]
  end

  # Dependencies can be Hex packages:
  #
  #   {:mydep, "~> 0.3.0"}
  #
  # Or git/path repositories:
  #
  #   {:mydep, git: "https://github.com/elixir-lang/mydep.git", tag: "0.1.0"}
  #
  # Type "mix help deps" for more examples and options
  defp deps do
    [
      {:earmark, "~> 1.0", only: :dev},
      {:ex_doc, "~> 0.13", only: :dev},
      {:excoveralls, "~> 0.5", only: :test},
      {:poison, "~> 2.2"}
    ]
  end

  defp package do
    [
      maintainers: ["Gary Fleshman"],
      licenses: ["MIT"],
      links: %{github: "https://github.com/garyf/apportion_ex"}
    ]
  end
end


lib/apportion.ex
defmodule Apportion do
  @moduledoc """
  Top level interface, or API, to proportionally distribute a quantity
  """

  alias Apportion.Algorithm.EqualProportions
  alias Apportion.Setup
  alias Apportion.Util

  @doc """
  Distributes a quantity proportionally among recipients per relative weights using the
    'equal proportions' algorithm

  ## Example
      iex> weights = %{a: 41, b: 32, c: 27}
      ...> Apportion.distribute(weights, 7)
      %{a: 3, b: 2, c: 2}
  """
  def distribute(weights, quantity, options \\ []) do
    portions = Setup.initial_portions(Map.keys(weights), options)
    portions_sum = Util.map_values_sum(portions)
    Setup.validate(quantity, portions_sum)
    next_step(weights, portions, portions_sum, quantity)
  end

  # recursively
  defp next_step(_, portions, portions_sum, quantity) when portions_sum == quantity do
    portions
  end
  defp next_step(weights, portions, portions_sum, quantity) do
    key = EqualProportions.recipient(weights, portions)
    next_portions = Map.merge(portions, %{key => Map.fetch!(portions, key) + 1})
    next_step(weights, next_portions, portions_sum + 1, quantity)
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :apportion, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:apportion, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/apportion/setup.ex
defmodule Apportion.Setup do
  @moduledoc """
  Setup functions
  """

  @doc """
  Determines initial portions for recipients per optional constraints

  ## Example
      iex> keys = [:a, :b, :c]
      ...> Apportion.Setup.initial_portions(keys, required_minimum: 2)
      %{a: 2, b: 2, c: 2}
  """
  def initial_portions(keys, options) do
    min = options[:required_minimum] || 0
    portions = Map.new(keys, fn x -> {x, min} end)
    include_prior_portions(portions, options[:prior_portions])
  end

  defp include_prior_portions(portions, nil), do: portions
  defp include_prior_portions(portions, prior_portions) do
    Map.merge(portions, prior_portions, fn (_k, v0, v1) -> Enum.max([v0, v1]) end)
  end

  def validate(quantity, portions_sum) when quantity < portions_sum do
    raise "Quantity less than minimum portions_sum"
  end
  def validate(_, _), do: nil
end


lib/apportion/algorithm/equal_proportions.ex
defmodule Apportion.Algorithm.EqualProportions do
  @moduledoc """
  Selects the next recipient

  see Balinski, M. and H. Young, The Quota Method of Apportionment,
    Amer. Math. Monthly 82 (1975) 701-730.
  """

  @doc """
  Selects the next recipient by sorting the equal proportions rank-index of the recipients

  ## Example
      iex> weights = %{a: 41, b: 32, c: 27}
      ...> portions = %{a: 4, b: 3, c: 2}
      ...> Apportion.Algorithm.EqualProportions.recipient(weights, portions)
      :c
  """
  def recipient(weights, portions) do
    {k, _v} = Enum.max_by(weights, fn {k, v} -> recipient_rank(v, Map.fetch!(portions, k)) end)
    k
  end

  defp recipient_rank(weight, portion) when portion == 0, do: zero_rank(weight)
  defp recipient_rank(weight, portion), do: weight / :math.sqrt(portion * (portion + 1))

  # protect from division by zero
  defp zero_rank(weight), do: weight * big_number

  def big_number, do: :math.pow(2, 30)
end


lib/apportion/util.ex
defmodule Apportion.Util do
  @moduledoc """
  Utility functions
  """

  @doc """
  Reads a JSON file and returns a map with atoms as keys

  ## Example
      iex> path_to_directory = "test/fixtures/balinski"
      ...> json_file_name = "table_9_populations.json"
      ...> Apportion.Util.file_read_json_to_map(path_to_directory, json_file_name)
      %{a: 9061, b: 7179, c: 5259, d: 3319, e: 1182}
  """
  def file_read_json_to_map(path_to_directory, json_file_name) do
    {:ok, map} = Path.join(path_to_directory, json_file_name)
      |> File.read!
      |> Poison.decode(keys: :atoms)
    map
  end

  @doc """
  Sums all of the values in a map

  ## Example
      iex> Apportion.Util.map_values_sum(%{a: 8, b: 2, c: 3})
      13
  """
  def map_values_sum(map) do
    Enum.reduce(map, 0, fn ({_k, v}, acc) -> v + acc end)
  end
end
